app/helpers/heatmap_helper.rb

```diff
@@ -71,25 +35,434 @@ def sample_taxons_dict(params)
+    FROM taxon_counts
+    -- warning!!! pipeline_runs may be missing!!!
+    LEFT OUTER JOIN pipeline_runs pr ON pipeline_run_id = pr.id
+    LEFT OUTER JOIN taxon_summaries ON
+      #{background_id.to_i}   = taxon_summaries.background_id   AND
+      taxon_counts.count_type = taxon_summaries.count_type      AND
+      taxon_counts.tax_level  = taxon_summaries.tax_level       AND
+      taxon_counts.tax_id     = taxon_summaries.tax_id
+    WHERE
+      pipeline_run_id IN (
+        SELECT MAX(id)
+        FROM pipeline_runs
+        WHERE sample_id IN (#{samples.pluck(:id).join(',')})
+        GROUP BY sample_id
+      )
+      AND genus_taxid != #{TaxonLineage::BLACKLIST_GENUS_ID}
+      AND count >= #{min_reads}
+      -- We need both types of counts for threshold filters
+      AND taxon_counts.count_type IN ('NT', 'NR')
+      #{categories_clause}
+      #{read_specificity_clause}
+      #{phage_clause}
+      AND #{zscore_sql} >= #{MINIMUM_ZSCORE_THRESHOLD}"
+
+    # This query:
+    # 1) assigns a rank to each row within a pipeline run
+    # 2) returns rows ranking <= num_results
+    # See http://www.sqlines.com/mysql/how-to/get_top_n_each_group
+    top_n_query = "
+      SELECT *
+      FROM (
+        SELECT
+          @rank := IF(@current_id = pipeline_run_id, @rank + 1, 1) AS rank,
+          @current_id := pipeline_run_id AS current_id,
+          a.*
+        FROM (
+          #{query}
+        ) a
+        ORDER BY
+          pipeline_run_id,
+          count_type = '#{sort[:count_type]}' DESC,
+          #{sort[:metric]} #{sort[:direction] == 'highest' ? 'DESC' : 'ASC'}
+      ) b
+      -- Overfetch by a factor of 4 to allow for post-SQL threshold filters
+      WHERE rank <= #{num_results * 4};
+    "
+
+    # TODO: (gdingle): how do we protect against SQL injection?
+    sql_results = TaxonCount.connection.select_all(top_n_query).to_hash
+
+    # organizing the results by pipeline_run_id
+    result_hash = {}
+
+    pipeline_run_ids = sql_results.map { |x| x['pipeline_run_id'] }
+    pipeline_runs = PipelineRun.where(id: pipeline_run_ids.uniq).includes([:sample])
+    pipeline_runs_by_id = Hash[pipeline_runs.map { |x| [x.id, x] }]
+
+    sql_results.each do |row|
+      pipeline_run_id = row["pipeline_run_id"]
+      if result_hash[pipeline_run_id]
+        pr = result_hash[pipeline_run_id]["pr"]
+      else
+        pr = pipeline_runs_by_id[pipeline_run_id]
+        result_hash[pipeline_run_id] = { "pr" => pr, "taxon_counts" => [] }
+      end
+      if pr.total_reads
+        z_max = ReportHelper::ZSCORE_MAX
+        z_min = ReportHelper::ZSCORE_MIN
+        row["zscore"] = z_max if row["zscore"] > z_max &&
+                                 row["zscore"] != ReportHelper::ZSCORE_WHEN_ABSENT_FROM_BACKGROUND
+        row["zscore"] = z_min if row["zscore"] < z_min
+        result_hash[pipeline_run_id]["taxon_counts"] << row
+      end
+    end
+    result_hash
+  end
+
+  def self.samples_taxons_details(
```

Comment:
this method and the rest below were moved with no change in logic from report helper